From 306cb151e9cc3f1fcb152b8f254fab5f3b03887a Mon Sep 17 00:00:00 2001
From: Raviteja Sunkara <raviteja@codeaurora.org>
Date: Mon, 20 May 2013 15:54:53 +0530
Subject: [PATCH 2/2] [RFC] Facilitate copying of query results.

Add a button to the query results screen which allows to copy
change urls(like https://gerrit-server-domain.com/change_number)
of all the search results to clipboard so that one can easily
share the results with others.

Note that only the links on the current page can be copied,
not all the search results. For others, you have to first load
next or previous search results and copy them.

Change-Id: Ic16d50e7a3194e70faa0f3094d29829ee6158440
---
 .../gerrit/client/changes/ChangeConstants.java     |    1 +
 .../client/changes/ChangeConstants.properties      |    1 +
 .../client/changes/PagedSingleListScreen.java      |   18 ++++++++++++++++++
 3 files changed, 20 insertions(+), 0 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java
index 33cdfbb..e3356de 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.java
@@ -171,6 +171,7 @@ public interface ChangeConstants extends Constants {
 
   String pagedChangeListPrev();
   String pagedChangeListNext();
+  String pagedCopySearchResults();
 
   String draftPatchSetLabel();
 
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties
index f5eb04b..30e9e5c 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeConstants.properties
@@ -148,6 +148,7 @@ buttonDeleteDraftPatchSet = Delete Draft Patch Set
 
 pagedChangeListPrev = &#x21e6;Prev
 pagedChangeListNext = Next&#x21e8;
+pagedCopySearchResults = Copy Search Results
 
 draftPatchSetLabel = (DRAFT)
 
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PagedSingleListScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PagedSingleListScreen.java
index c21c68d..471a918 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PagedSingleListScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/PagedSingleListScreen.java
@@ -16,6 +16,7 @@ package com.google.gerrit.client.changes;
 
 import com.google.gerrit.client.Gerrit;
 import com.google.gerrit.client.rpc.ScreenLoadCallback;
+import com.google.gerrit.client.ui.ChangeLink;
 import com.google.gerrit.client.ui.Hyperlink;
 import com.google.gerrit.client.ui.Screen;
 import com.google.gerrit.reviewdb.client.AccountGeneralPreferences;
@@ -23,6 +24,7 @@ import com.google.gwt.event.dom.client.KeyPressEvent;
 import com.google.gwt.user.client.History;
 import com.google.gwt.user.client.rpc.AsyncCallback;
 import com.google.gwt.user.client.ui.HorizontalPanel;
+import com.google.gwtexpui.clippy.client.CopyableLabel;
 import com.google.gwtexpui.globalkey.client.KeyCommand;
 
 public abstract class PagedSingleListScreen extends Screen {
@@ -35,6 +37,7 @@ public abstract class PagedSingleListScreen extends Screen {
   protected Hyperlink prev;
   protected Hyperlink next;
   protected ChangeList changes;
+  private CopyableLabel copySearchResult;
 
   protected final String anchorPrefix;
   protected boolean useLoadPrev;
@@ -76,6 +79,12 @@ public abstract class PagedSingleListScreen extends Screen {
     section = new ChangeTable2.Section();
     table.addSection(section);
     table.setSavePointerId(anchorPrefix);
+
+    HorizontalPanel copySearchResultPanel = new HorizontalPanel();
+    copySearchResultPanel.add(new Label(Util.C.pagedCopySearchResults()));
+    copySearchResultPanel.add(copySearchResult = new CopyableLabel("", false));
+    setTitleFarEast(copySearchResultPanel);
+
     add(table);
 
     final HorizontalPanel buttons = new HorizontalPanel();
@@ -133,9 +142,18 @@ public abstract class PagedSingleListScreen extends Screen {
     }
     table.updateColumnsForLabels(result);
     section.display(result);
+    updateCopySearchResult();
     table.finishDisplay();
   }
 
+  protected void updateCopySearchResult() {
+    String links = "";
+    for(int i = 0; i < changes.size(); i++){
+      links += ChangeLink.permalink(changes.get(i).legacy_id()) + "\n";
+    }
+    copySearchResult.setText(links);
+  }
+
   private static final class DoLinkCommand extends KeyCommand {
     private final Hyperlink link;
 
-- 
1.7.8.3

